File: source/115.不同的子序列.cpp
/*
 * @lc app=leetcode.cn id=115 lang=cpp
 *
 * [115] 不同的子序列
 */
#include "cpp_includes.h"
// @lc code=start
class Solution {
  public:
    int numDistinct(string s, string t) {
        int m = t.length();
        int n = s.length();

        vector<vector<long>> f(m + 1, vector<long>(n + 1, 0));

        for (int i = 0; i <= n; i++) {
            f[0][i] = 1;
        }

        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                if (t[i - 1] == s[j - 1]) {
                    f[i][j] = f[i][j - 1] + f[i - 1][j - 1];
                } else {
                    f[i][j] = f[i][j - 1];
                }
            }
        }
        return f[m][n];
    }
};
// @lc code=end

File: source/115.不同的子序列.py
#
# @lc app=leetcode.cn id=115 lang=python3
#
# [115] 不同的子序列
#

# @lc code=start
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        n, m = len(s), len(t)
        dp = [[0]*(n+1) for _ in range(m+1)]
        for j in range(n+1):
            dp[0][j] = 1
        for i in range(1, m+1):
            for j in range(1, n+1):
                if t[i-1] == s[j-1]:
                    dp[i][j] = dp[i-1][j-1] + dp[i][j-1]
                else:
                    dp[i][j] = dp[i][j-1]

        return dp[-1][-1]

# @lc code=end
if __name__ == '__main__':
    S = Solution()
    S.numDistinct("rabbbit", "rabbit")
